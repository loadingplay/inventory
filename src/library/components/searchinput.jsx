import React from 'react';
import PropTypes from 'prop-types';


function debounce(func, wait, immediate) {
  var timeout;
  return function () {
    var context = this, args = arguments;
    var later = function () {
      timeout = null;
      if (!immediate) func.apply(context, args);
    };
    var callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) func.apply(context, args);
  };
}


export class SearchInput extends React.Component {

  static propTypes = {
    onSearch: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.state = { value: '' }
    this.textInput = React.createRef();

    this.handleSearch = debounce(props.onSearch, 500);
  }

  componentDidMount = () => {
    document.addEventListener('keydown', this.handleKeyDown);
  }

  componentWillUnmount = () => {
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  handleChange = (event) => {
    this.setState({
      value: event.target.value
    });

    this.handleSearch(event.target.value);
  }

  handleKeyDown = (e) => {
    if (
      e.key !== "Enter" &&
      e.key !== "Tab" &&
      e.key !== "Shift"
    ) {
      this.textInput.current.focus();
    }
  }

  render = () => {
    return (
      <div className="form-group row">
        <label htmlFor="search" className="col-sm-2 col-form-label">SKU</label>
        <div className="col-sm-10">
          <input
            type="text"
            ref={this.textInput}
            placeholder="ingrese sku"
            className="form-control"
            id="search"
            value={this.state.value}
            onChange={this.handleChange}
          />
        </div>
      </div>
    );
  }
}
