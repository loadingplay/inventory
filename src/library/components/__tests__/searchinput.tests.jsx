import React from 'react';
import { shallow } from 'enzyme';
import { SearchInput } from "../searchinput";

describe("SearchInput", () => {
  test("it should render a seach box", () => {
    const wrapper = shallow(<SearchInput />);

    expect(wrapper).toMatchSnapshot("default");
  });

  test("it should change search type", () => {
    const onTypeChange = jest.fn();
    const wrapper = shallow(<SearchInput onTypeChange={onTypeChange} />);

    expect(wrapper.instance().state.type).toEqual("search");
    expect(onTypeChange).toHaveBeenCalledTimes(0);
    wrapper.instance().changeType();
    expect(wrapper.instance().state.type).toEqual("barcode");
    expect(onTypeChange).toHaveBeenCalledTimes(1);
    wrapper.instance().changeType();
    expect(wrapper.instance().state.type).toEqual("search");
    expect(onTypeChange).toHaveBeenCalledTimes(2);
  });

  test("it should handle when keydown", () => {
    const wrapper = shallow(<SearchInput />);
    const e = { key: "Enter", target: { select: jest.fn() } };

    wrapper.instance().textInput = { current: { focus: jest.fn() } };

    expect(wrapper.instance().textInput.current.focus).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown({ key: "t" });
    expect(wrapper.instance().textInput.current.focus).toHaveBeenCalledTimes(1);

    expect(e.target.select).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown(e);
    expect(e.target.select).toHaveBeenCalledTimes(1);
  });

  test("it should handle changes", () => {
    const wrapper = shallow(<SearchInput />);

    wrapper.instance().handleSearch = jest.fn();

    expect(wrapper.instance().state.value).toEqual("");
    wrapper.instance().handleChange({ target: { value: "test" } });
    expect(wrapper.instance().state.value).toEqual("test");
    expect(wrapper.instance().handleSearch).toHaveBeenCalledWith("test");
  });

});
