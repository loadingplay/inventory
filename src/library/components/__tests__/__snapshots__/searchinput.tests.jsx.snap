// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput it should render a seach box: default 1`] = `
<div
  className="form-group row no-gutters"
>
  <div
    className="col"
  >
    <div
      className="input-group mb-3"
    >
      <input
        className="form-control"
        id="search"
        onChange={[Function]}
        placeholder="ingresar bÃºsqueda"
        type="text"
        value=""
      />
      <div
        className="input-group-append"
      >
        <span
          className="input-group-text"
        >
          <i
            className="fas fa-search"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="col-auto"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-sync"
      />
    </button>
  </div>
</div>
`;
